Podcast/Player.swift

```diff
@@ -9,59 +9,189 @@
+    case paused
+    case failed
+}
+
+/// Protocol for the singleton Player to observe changes in the Player's state and time changes
+protocol PlayerDelegate {
+    /// Called when the Player's state changes. Override to observe when the state is changed.
+    func playerDidChangeState() -> Void
+    /// Called when the Player has made playback progress. Override to observe when the Player makes progress on playback.
+    func playerDidUpdateTime() -> Void
+}
+
+/// Handles playback of podcasts from URLs
+class Player: NSObject {
     
-    init(fileURL: URL) {
-        self.fileURL = fileURL
+    static let sharedInstance = Player()
+    private override init() {
+        playerStatus = .empty
+        shouldAutoPlay = false
         super.init()
     }
     
-    var isPlaying: Bool {
-        return player?.isPlaying ?? false
+    var playerStatus: PlayerStatus {
+        didSet {
+            if oldValue != playerStatus {
+                self.delegate?.playerDidChangeState()
+            }
+        }
     }
     
-    func prepareToPlay() {
-        if player == nil {
-            if fileURL.isFileURL {
-                player = try? AVAudioPlayer(contentsOf: fileURL as URL)
-                player?.prepareToPlay()
-            } else if currentTask == nil {
-                let request = URLRequest(url: fileURL as URL, cachePolicy:  .returnCacheDataElseLoad, timeoutInterval: 15)
-                currentTask = URLSession.shared.dataTask(with: request) { [weak self] data, response, _ -> Void in
-                    guard let s = self else { return }
-                    guard let data = data else { return }
-                    
-                    s.player = try? AVAudioPlayer(data: data)
-                    s.player?.prepareToPlay()
-                    s.currentTask = nil
-                    s.player?.play()
-                }
-                currentTask?.resume()
+    var delegate: PlayerDelegate?
+    private var shouldAutoPlay: Bool
+    private var playerContext: UnsafeMutableRawPointer?
+    private var currentAVPlayer: AVPlayer? {
+        didSet {
+            oldValue?.pause()
+        }
+    }
+    private var currentURL: URL? {
+        didSet {
+            if oldValue != currentURL {
+                prepareToPlay(url: currentURL!)
             }
         }
     }
     
+    /// Prepares the player to play a new track from a URL. If the Player is in the .preparingToPlay state, this has no effect on current playback.
+    /// - Parameters:
+    ///   - url: The url to fetch and play from.
+    func prepareToPlay(url: URL) {
+        if currentURL != url || playerStatus != .preparingToPlay {
+            // create a new player if this is a new URL or we aren't currently preparing a URL
+            playerStatus = .preparingToPlay
+            currentAVPlayer = AVPlayer(url: url)
+            currentAVPlayer?.addObserver(self, forKeyPath: #keyPath(AVPlayer.status), options: [.old, .new], context: &playerContext)
```

Comment:
oops 😆 I now add an observer in `play()` and remove it in `pause()` since we don't need to observe the time when it's paused. also added a `deinit` to remove it there